packages/grid/_modules_/grid/hooks/features/columnPinning/useGridColumnPinning.tsx

```diff
@@ -0,0 +1,304 @@
+import { GridComponentProps } from '../../../GridComponentProps';
+import {
+  visibleGridColumnsSelector,
+  gridColumnsMetaSelector,
+} from '../columns/gridColumnsSelector';
+import { GridCellIndexCoordinates } from '../../../models/gridCell';
+import { GridPreProcessingGroup } from '../../core/preProcessing';
+import { useGridApiEventHandler } from '../../utils/useGridApiEventHandler';
+import { GridEvents } from '../../../constants/eventsConstants';
+import { GridCellParams } from '../../../models/params/gridCellParams';
+import { gridClasses } from '../../../gridClasses';
+import { useGridRegisterPreProcessor } from '../../core/preProcessing/useGridRegisterPreProcessor';
+import { GridColumnPinningMenuItems } from '../../../components/menu/columnMenu/GridColumnPinningMenuItems';
+import { GridColumns } from '../../../models/colDef/gridColDef';
+import { useGridApiMethod } from '../../utils/useGridApiMethod';
+import { GridColumnPinningApi, GridPinnedPosition } from '../../../models/api/gridColumnPinningApi';
+import { gridPinnedColumnsSelector } from './columnPinningSelector';
+import { useGridStateInit } from '../../utils/useGridStateInit';
+import { useGridState } from '../../utils/useGridState';
+import { useGridSelector } from '../../utils/useGridSelector';
+import { filterColumns } from '../../../../../x-data-grid-pro/src/DataGridProVirtualScroller';
+
+const Divider = () => <MuiDivider onClick={(event) => event.stopPropagation()} />;
+
+export const useGridColumnPinning = (
+  apiRef: GridApiRef,
+  props: Pick<
+    GridComponentProps,
+    'initialState' | 'disableColumnPinning' | 'pinnedColumns' | 'onPinnedColumnsChange'
+  >,
+): void => {
+  useGridStateInit(apiRef, (state) => ({
+    ...state,
+    pinnedColumns: {
+      left: !props.disableColumnPinning ? props.initialState?.pinnedColumns?.left : undefined,
+      right: !props.disableColumnPinning ? props.initialState?.pinnedColumns?.right : undefined,
+    },
+  }));
+  const [, setGridState, forceUpdate] = useGridState(apiRef);
+  const pinnedColumns = useGridSelector(apiRef, gridPinnedColumnsSelector);
+
+  // Each visible row (not to be confused with a filter result) is composed of a central .MuiDataGrid-row element
+  // and up to two additional .MuiDataGrid-row's, one for the columns pinned to the left and another
+  // for those on the right side. When hovering any of these elements, the :hover styles are applied only to
+  // the row element that was actually hovered, not its additional siblings. To make it look like a contiguous row,
+  // this method adds/removes the .Mui-hovered class to all of the row elements inside one visible row.
+  const updateHoveredClassOnSiblingRows = React.useCallback(
+    (event: React.MouseEvent<HTMLDivElement>) => {
+      if (!Array.isArray(pinnedColumns.left) && !Array.isArray(pinnedColumns.right)) {
+        return;
+      }
+
+      if (pinnedColumns.left!.length === 0 && pinnedColumns.right!.length === 0) {
```

Comment:
This one crashes if `left` is defined but not `right` or the opposite.
You could replace both `if` statement with

```suggestion
      if ((pinnedColumns.left?.length ?? 0 + pinnedColumns.right?.length ?? 0) === 0) {
```

Reply:
Fixed.